src/citrine/resources/table_config.py

```diff
@@ -237,6 +238,93 @@ def add_all_ingredients(self, *,
         new_config.version_uid = copy(self.version_uid)
         return new_config
 
+    def add_all_ingredients_in_output(self, *,
+                                      process_templates: List[LinkByUID],
+                                      project,
+                                      quantity_dimension: IngredientQuantityDimension,
+                                      scope: str = CITRINE_SCOPE,
+                                      unit: Optional[str] = None
+                                      ):
+        """[ALPHA] Add variables and columns for all of the possible ingredients in a list of
+        processes.
+
+        For each allowed ingredient name in the union of all passed process templates there is a
+        column for the id of the ingredient and a column for the quantity of the ingredient.
+        Columns are filled with the "InOutput" method halting at any of the passed process
+        templates. If the quantities are given in absolute amounts then there is also a column for
+        units.
+
+        Parameters
+        ------------
+        process_templates: List[LinkByUID]
+            registered process templates from which to pull allowed ingredeintes and at which to
+            halt searching
+        project: Project
+            a project that has access to the process template
+        quantity_dimension: IngredientQuantityDimension
+            the dimension in which to report ingredient quantities
+        scope: Optional[str]
+            the scope for which to get ingredient ids (default is Citrine scope, 'id')
+        unit: Optional[str]
+            the units for the quantity, if selecting Absolute Quantity
+
+        """
+        dimension_display = {
+            IngredientQuantityDimension.ABSOLUTE: "absolute quantity",
+            IngredientQuantityDimension.MASS: "mass fraction",
+            IngredientQuantityDimension.VOLUME: "volume fraction",
+            IngredientQuantityDimension.NUMBER: "number fraction"
+        }
+        union_allowed_names = []
+        for process_template_link in process_templates:
+            process: ProcessTemplate = project.process_templates.get(process_template_link)
+            if not process.allowed_names:
+                raise RuntimeError(
+                    "Cannot add ingredients for process template \'{}\' because it has no defined "
+                    "ingredients (allowed_names is not defined).".format(process.name))
```

Comment:
```suggestion
                    f"Cannot add ingredients for process template '{process.name}' "
                    "because it has no defined ingredients (allowed_names is not defined)"
```
If you are so inclined, we changed the minimum python version so we now allow for format strings.

Reply:
I prefer f strings as well